compiler/src/dotty/tools/dotc/transform/localopt/RemoveUnnecessaryNullChecks.scala

```diff
@@ -6,104 +6,120 @@ import core.Contexts.Context
+              case t @ NullCheck(sym) if nullness.contains(sym) => Literal(Constant(nullness(sym)))
+              case t => t
+            }
+          }
+        }.transform(tree)
+
+    {
+      case blk @ Block(stats, expr) => 
+        // (symbol -> is null) elements for which we don't have informations aren't in the map
+        val nullness = mutable.Map[Symbol, Boolean]()
+
+        // map statements and propagate nullness
+        val newStats = stats.mapConserve {
+          case t: ValDef if !isVar(t.symbol) => 
+            if (isAlwaysNull(t.rhs)) nullness += t.symbol -> true
+            if (isNeverNull(t.rhs)) nullness += t.symbol -> false
+            t
+
+          // throw NPE if id is null
+          case t @ Apply(Select(id: Ident, _), _) if !isVar(id.symbol) => 
+            nullness += id.symbol -> false
+            t
+
+          case t => 
+            transform(t, nullness)
+        }
 
-          case t: Ident if !t.symbol.owner.isClass =>
-            checkGood.update(target, checkGood.getOrElse(target, Set.empty) + t.symbol)
-            true
+        val newExpr = transform(expr, nullness)
 
-          case t: Apply if !t.symbol.owner.isClass =>
-            checkGood.update(target, checkGood.getOrElse(target, Set.empty) + t.symbol)
-            true
+        if ((newStats eq stats) && (newExpr eq expr)) blk 
+        else Block(newStats, newExpr)
 
-          case t: Typed =>
-            checkNonNull(t.expr, target)
+      // if (x == null)
+      case br @ If(cond @ NullCheck(sym), thenp, elsep) => 
+        val newThen = transform(thenp, mutable.Map(sym -> true))
+        val newElse = transform(elsep, mutable.Map(sym -> false))
+        if ((newThen eq thenp) && (newElse eq elsep)) br
+        else If(cond, newThen, newElse)
 
-          case _ => t.tpe.isNotNull
+      // if (x != null)
+      case br @ If(cond @ Select(NullCheck(sym), _), thenp, elsep) if cond.symbol eq defn.Boolean_! =>
+        val newThen = transform(thenp, mutable.Map(sym -> false))
+        val newElse = transform(elsep, mutable.Map(sym -> true))
+        if ((newThen eq thenp) && (newElse eq elsep)) br
+        else If(cond, newThen, newElse)
 
-        }
-        if (checkNonNull(vd.rhs, vd.symbol))
-          initializedVals += vd.symbol
-      }
-    case t: Tree =>
+      case t => t
+    }
   }
 
 
-  def transformer(implicit ctx: Context): Tree => Tree = {
-    def isNullLiteral(tree: Tree) = tree match {
-      case literal: Literal =>
-        literal.const.tag == NullTag
-      case _ => false
+  private def isNeverNull(tree: Tree)(implicit ctx: Context): Boolean =
+    tree match {
+      case Block(_, expr) => isNeverNull(expr)
+      case If(_, th, el) => isNeverNull(th) && isNeverNull(el)
+      case t: Typed => isNeverNull(t.expr)
+      case t: Literal => t.const.tag != NullTag
+      case _: This => true
+      case _: New => true
+      case t: Apply if t.symbol.isPrimaryConstructor => true
```

Comment:
why `Primary`?

Reply:
It's taken from [here](https://github.com/lampepfl/dotty/blob/9de3905ca9e287d42ef293025d0ff3369b7d8f12/compiler/src/dotty/tools/dotc/transform/localopt/RemoveUnnecessaryNullChecks.scala#L60).